### DISCUSSION TITLE
Extend LDAP Auth to take a priority list for group->role mapping

### DISCUSSION THREAD

**Discussion by marcules** [directus/directus#12844] [Feature Requests]
Hi,

I understand that directus currently has a 1:1 mapping for users and roles and wanted to ask if it would be an acceptable / good idea to have a prioritized list of ldap groups that will be mapped for a user.

Let's say you have the groups "MY_LDAP_GROUP_APPLICATION_NAME_TEST_USER" and "MY_LDAP_GROUP_APPLICATION_NAME_TEST_CMS_ADMIN" in LDAP and want to enable a user in the first group to access the CMS in a read-only mode and the user in the second group to access the CMS in rw-mode - then I would create two roles with the same name and a user that has both groups would get the alphabetically sorted  (or is it random?)  first role.

I think it's an OK compromise to have a list of roles for this provider so that it is first checked if the user is in the first item, then so on, and assign the role that is first found - then the restriction to assign a user to a signular group in ldap would be lifted and there would be a predictable way to assign the correct role out of a bunch of "allowed groups".

@aidenfoxx @rijkvanzanten what do you think? Can/Should I create a PR for that?

LG Marc

**Comment by aidenfoxx** [comment]
The order isn't random, but is based on whatever order your roles are in the database (if I remember correctly). So a short term solution would be to have your "Admin" role at the top of the table in your database.

That being said, all this will likely be solved if Directus adds multi-role users which is currently on the roadmap for Q2 2022? https://directus.io/road-map/

**Comment by marcules** [comment]
Do you think it would be appropriate to have a "quick win" with a predefined group->role map before that is implemented, to have a explicitly defined behaviour, or is that out of scope?

I think the described fix would at least allow for a temporary mapping that does not depend on a complete overhaul (and could be scoped into 9.9.x) - wdyt?

**Reply by aidenfoxx** [reply]
I think the quick win is sorting the roles in the DB in order of priority, but I am also lazy so there is that. I'm also against adding code that we know will be depreciated in the near future. Personally I would keep my fingers crossed for multi-role users in Q2. 🤞

**Reply by marcules** [reply]
@aidenfoxx sorry to be a bother, but as far as I can see the order is based on the role ID - which I am not able to influence?

https://github.com/directus/directus/blob/main/api/src/auth/drivers/ldap.ts#L251-L260

Could you expand on the "sorting the roles in the DB in order of priority" < how would I do that? I don't think I can influence the role ID through CLI or REST client, right?

![image](https://user-images.githubusercontent.com/2453108/164004132-19a0086d-4ebb-47ca-a5f7-e1366974cf4b.png)

**Reply by marcules** [reply]
What I mean is - I cannot set a "default order" in a SQL table, right? Or am I misunderstanding the priority / sorting you mean?

**Reply by aidenfoxx** [reply]
I think it depends on your platform, but looking into it, most SQL platforms the order does seem to be random. I'm coming from Oracle world where I believe everything is ordered by an in visible row ID when an order isn't specified.

In that case I guess the only way to currently guarantee a role is to make sure users aren't in multiple Directus groups in AD.

**Reply by marcules** [reply]
Would you accept a PR where I sort the returned groups by a user-provided priority list first?

something like:

```js
if (userGroups.length) {
let userRoles = await this.knex
	.select('id', 'name')
	.from('directus_roles')
	.whereRaw(`LOWER(??) IN (${userGroups.map(() => '?')})`, [
		'name',
		...userGroups.map((group) => group.toLowerCase()),
	]);

userRoles.sort((roleA, roleB) => groupPrios.indexOf(roleA.name.toLowerCase()) - groupPrios.indexOf(roleB.name.toLowerCase()));

userRole = userRoles.length > 0 ? userRoles[0] : null;
}
```

Where `groupPrios` is a list of group names in the prioritization order (first in list has higher order).

**Reply by aidenfoxx** [reply]
@marcules That would be up to @rijkvanzanten to decide. There would be no need for this once the multi-role user support is added, and it's not too hard to work around currently by ensuring that your AD users only belong to 1 Directus group. For me it doesn't feel super necessary. 🤷

**Reply by rijkvanzanten** [reply]
I don't know enough about LDAP and common setups to make any judgement on how common this is, but it feels a little 80/20 territory to me 🤔  I do agree with @aidenfoxx above though in that I definitely don't want to add any temporary functionality that we know is going to be deprecated once multi-role is a thing. If Directus' multi-role capabilities solves for the same problem, I'd prefer focussing the efforts there 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
